### PR TITLE
Added EventHandlerEntityBundle.

### PR DESCRIPTION
**Pull request by milkovsky**

### CODE REVIEW

File: src/EventHandler/EventHandlerBase.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\rules\Core\RulesConfigurableEventHandlerInterface;
+use Drupal\rules\Core\RulesDefaultEventHandler;
+use Symfony\Component\EventDispatcher\Event;
+
+/**
+ * Base class for event handler.
+ */
+abstract class EventHandlerBase extends RulesDefaultEventHandler implements RulesConfigurableEventHandlerInterface {
```
Comment (fago) [inline src/EventHandler/EventHandlerBase.php]: Should be ConfigureableEventHandlerBase  - as that what it implements. same for comment.

File: src/EventHandler/EventHandlerBase.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  public static function determineQualifiedEvents(Event $event, $event_name, array &$event_definition) {
+    return array();
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function summary() {
+    return '';
```
Comment (fago) [inline src/EventHandler/EventHandlerBase.php]: this default is not useful -> leave it abstract

File: src/EventHandler/EventHandlerBase.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+   *
+   * @var array
+   */
+  protected $configuration = array();
+
+  /**
+   * @inheritdoc
+   */
+  public static function determineQualifiedEvents(Event $event, $event_name, array &$event_definition) {
+    return array();
```
Comment (fago) [inline src/EventHandler/EventHandlerBase.php]: this default is not useful -> leave it abstract

File: src/EventHandler/EventHandlerBase.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  public function summary() {
+    return '';
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    return $form;
```
Comment (fago) [inline src/EventHandler/EventHandlerBase.php]: this default is not useful -> leave it abstract

File: src/EventHandler/EventHandlerBase.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  public function defaultConfiguration() {
+    return array();
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function getEventNameSuffix() {
+    return '';
```
Comment (fago) [inline src/EventHandler/EventHandlerBase.php]: this default is not useful -> leave it abstract

File: src/EventHandler/EventHandlerEntityBundle.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\rules\EventHandler\EventHandlerEntityBundle.
+ */
+
+namespace Drupal\rules\EventHandler;
+use Symfony\Component\EventDispatcher\Event;
```
Comment (fago) [inline src/EventHandler/EventHandlerEntityBundle.php]: missing NL before use

File: src/EventHandler/EventHandlerEntityBundle.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+  /**
+   * @inheritdoc
+   */
+  public static function determineQualifiedEvents(Event $event, $event_name, array &$event_definition) {
+    $events_suffixes = [];
+    if ($event instanceof \Drupal\rules\Event\EntityEvent) {
+      $events[] = $event->getSubject()->bundle();
+    }
+    return $events_suffixes;
```
Comment (fago) [inline src/EventHandler/EventHandlerEntityBundle.php]: no change on variable?

File: src/EventHandler/EventHandlerBase.php (hunk 3 above)
Comment (milkovsky) [inline src/EventHandler/EventHandlerBase.php]: should I add 'abstract' ?

File: tests/src/Kernel/EventHandlerTest.php [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    // Create rule with the 'rules_entity_presave:node--page' event.
+    $rule = $this->expressionManager->createRule();
+    $rule->addAction('rules_test_log');
+    $config_entity = $this->storage->create([
+      'id' => 'test_rule',
+      'expression_id' => 'rules_rule',
+      'event' => 'rules_entity_presave:node--page',
+      'configuration' => $rule->getConfiguration(),
+    ]);
+    $config_entity->save();
```
Comment (milkovsky) [inline tests/src/Kernel/EventHandlerTest.php]: Once it is saved `ReactionRuleStorage::save()`, rules creates a listener to `rules_entity_presave:node--page`, not to `rules_entity_presave:node` at the line: 

`$this->stateService->set('rules.registered_events', $this->getRegisteredEvents());`

So `rules_entity_presave:node` is not listerned.

How could we fix it?

Comment (fago) [inline tests/src/Kernel/EventHandlerTest.php]: We need to adapt this code and get the event base name from the qualified event name before adding the listener. The D7 code will work :

```

/**
 * Returns the base name of a configured event name.
 *
 * For a configured event name like node_view--article the base event name
 * node_view is returned.
 *
 * @return string
 *   The event base name.
 */
function rules_get_event_base_name($event_name) {
  // Cut off any suffix from a configured event name.
  if (strpos($event_name, '--') !== FALSE) {
    $parts = explode('--', $event_name, 2);
    return $parts[0];
  }
  return $event_name;
}
```

Not sure where to put it best, I'd suggest adding to RulesEventManage

File: tests/src/Kernel/EventHandlerTest.php [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    $rule = $this->expressionManager->createRule();
+    $config_entity = $this->storage->create([
+      'id' => 'test_rule',
+      'expression_id' => 'rules_rule',
+      'event' => 'rules_entity_presave:node-–page',
+      'configuration' => $rule->getConfiguration(),
+    ]);
+    $config_entity->save();
+
+    // @todo Add integrity check that node.field_integer is detected by Rules.
```
Comment (milkovsky) [inline tests/src/Kernel/EventHandlerTest.php]: testEntityBundleHandlerExecution() already covers node.field_integer detection.
It adds ->map('message', 'node.field_integer.0.value') to rules log.

Sould I remove testEntityBundleHandlerConfiguration() and rename testEntityBundleHandlerExecution to testEntityBundleHandler?

File: src/Core/RulesEventManager.php [hunk 10]
```diff
@@ -60,4 +60,22 @@ public function processDefinition(&$definition, $plugin_id) {
...
+  /**
+   * Returns the base name of a configured event name.
+   *
+   * For a configured event name like node_view--article the base event name
+   * node_view is returned.
+   *
+   * @return string
+   *   The event base name.
+   */
+  static public function rulesGetEventBaseName($event_name) {
```
Comment (milkovsky) [inline src/Core/RulesEventManager.php]: static, ok?

Comment (fago) [inline src/Core/RulesEventManager.php]: no, statics should be avoided if possible. the manager is instantiated anyway, so do not make it static

File: src/Core/RulesEventManager.php [hunk 11]
```diff
@@ -60,4 +60,22 @@ public function processDefinition(&$definition, $plugin_id) {
     }
   }
 
+  /**
+   * Returns the base name of a configured event name.
```
Comment (fago) [inline src/Core/RulesEventManager.php]: minor: Gets

File: src/Core/RulesEventManager.php [hunk 12]
```diff
@@ -60,4 +60,22 @@ public function processDefinition(&$definition, $plugin_id) {
...
+   *
+   * For a configured event name like node_view--article the base event name
```
Comment (fago) [inline src/Core/RulesEventManager.php]: oudated examples, check configurable event interface and also add an @see to it as this holds the related docs

File: src/Core/RulesEventManager.php (hunk 10 above)
Comment (fago) [inline src/Core/RulesEventManager.php]: method name is bad:
- no rules prefix needed
- methods should start with a verb
  -> getEventBaseName()

File: src/Entity/ReactionRuleStorage.php [hunk 13]
```diff
@@ -84,6 +85,7 @@ protected function getRegisteredEvents() {
     $events = [];
     foreach ($this->loadMultiple() as $rules_config) {
       $event = $rules_config->getEvent();
+      $event = RulesEventManager::rulesGetEventBaseName($event);
```
Comment (fago) [inline src/Entity/ReactionRuleStorage.php]: You need to get the rules event manager service here. If not done yet, inject it first.

File: src/EventHandler/ConfigurableEventHandlerBase.php [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  public static function determineQualifiedEvents(Event $event, $event_name, array &$event_definition) {
+    // Nothing to do by default.
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function summary() {
+    // Nothing to do by default.
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerBase.php]: needs to be removed

File: src/EventHandler/ConfigurableEventHandlerBase.php [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+  /**
+   * The event configuration.
+   *
+   * @var array
+   */
+  protected $configuration = array();
+
+  /**
+   * @inheritdoc
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerBase.php]: wrong doc formatting, check other classes. correct below also

File: src/EventHandler/ConfigurableEventHandlerBase.php [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  public function summary() {
+    // Nothing to do by default.
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerBase.php]: as reviewed already, remove useless examples. check old PR - this was not addressed

File: src/EventHandler/ConfigurableEventHandlerEntityBundle.php [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+use Symfony\Component\EventDispatcher\Event;
+
+/**
+ * Exposes the bundle of an entity as event setting.
+ */
+class ConfigurableEventHandlerEntityBundle extends ConfigurableEventHandlerBase {
+
+  /**
+   * @inheritdoc
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php]: wrong format

File: tests/src/Kernel/EventHandlerTest.php (hunk 9 above)
Comment (fago) [inline tests/src/Kernel/EventHandlerTest.php]: merging methods seems fine, but we need to keep the todo to test this via an integrity check also

File: tests/src/Kernel/EventHandlerTest.php [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    $rule1 = $this->expressionManager->createRule();
+    $rule1->addAction('rules_test_log',
+      ContextConfig::create()
+        ->map('message', 'node.field_integer.0.value')
+    );
+    $config_entity1 = $this->storage->create([
+      'id' => 'test_rule1',
+      'expression_id' => 'rules_rule',
+      'event' => 'rules_entity_presave:node--page',
+      'configuration' => $rule1->getConfiguration(),
```
Comment (fago) [inline tests/src/Kernel/EventHandlerTest.php]: ok, but its better to use setters instead. only pass to create() what's absolutely necessary (the bundle) and use setters else.

File: src/Entity/ReactionRuleStorage.php (hunk 13 above)
Comment (milkovsky) [inline src/Entity/ReactionRuleStorage.php]: Not sure what you meant with 'inject'... I implemented as ReactionRuleStorage::eventManager().

File: src/EventHandler/ConfigurableEventHandlerBase.php (hunk 14 above)
Comment (milkovsky) [inline src/EventHandler/ConfigurableEventHandlerBase.php]: Method or comment? If I remove the method I have to implement it in the ConfigurableEventHandlerEntityBundle. What to put in there?

File: src/EventHandler/ConfigurableEventHandlerBase.php (hunk 16 above)
Comment (milkovsky) [inline src/EventHandler/ConfigurableEventHandlerBase.php]: there were no Handler classes in the old PR. The same Q as above: If I remove the method I have to implement it in the ConfigurableEventHandlerEntityBundle. What to put in there?

**Comment by milkovsky** [comment]
> fago: as reviewed already, remove useless examples. check old PR - this was not addressed

there were no Handler classes in the old PR.
If I remove the methods I have to implement them in the ConfigurableEventHandlerEntityBundle. What to put in there?

File: src/Core/RulesEventManager.php [hunk 19]
```diff
@@ -60,4 +60,24 @@ public function processDefinition(&$definition, $plugin_id) {
...
+  /**
+   * Gets the base name of a configured event name.
+   *
+   * For a configured event name like {EVENT_NAME}--{SUFFIX}, the base event
+   * name {EVENT_NAME} is returned.
+   *
+   * @return string
+   *   The event base name.
+   *
+   * @see RulesConfigurableEventHandlerInterface::getEventNameSuffix()
```
Comment (fago) [inline src/Core/RulesEventManager.php]: must use fully qualified class name on references

File: src/Entity/ReactionRuleStorage.php [hunk 20]
```diff
@@ -40,6 +41,13 @@ class ReactionRuleStorage extends ConfigEntityStorage {
   protected $drupalKernel;
 
   /**
+   * The event manager.
+   *
+   * @var RulesEventManager
```
Comment (fago) [inline src/Entity/ReactionRuleStorage.php]: missing namespace.

File: src/Entity/ReactionRuleStorage.php [hunk 21]
```diff
@@ -61,6 +69,19 @@ public function __construct(EntityTypeInterface $entity_type, ConfigFactoryInter
   }
 
   /**
+   * Gets the event manager.
+   *
+   * @return RulesEventManager
```
Comment (fago) [inline src/Entity/ReactionRuleStorage.php]: missing namespace.

File: src/EventHandler/ConfigurableEventHandlerBase.php [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+   *
+   * @var array
+   */
+  protected $configuration = [];
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function determineQualifiedEvents(Event $event, $event_name, array &$event_definition) {
+    // Nothing to do by default.
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerBase.php]: not useful default method, implementation must provide this - should be removed

File: src/EventHandler/ConfigurableEventHandlerBase.php [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function summary() {
+    // Nothing to do by default.
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerBase.php]: not useful default method, implementation must provide this - should be removed

File: src/EventHandler/ConfigurableEventHandlerBase.php [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    // Nothing to do by default.
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerBase.php]: not useful default method, implementation must provide this - should be removed

File: src/EventHandler/ConfigurableEventHandlerBase.php [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  public function defaultConfiguration() {
+    return [];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getEventNameSuffix() {
+    // Nothing to do by default.
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerBase.php]: not useful default method, implementation must provide this - should be removed

File: tests/src/Kernel/EventHandlerTest.php [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+use Drupal\field\Entity\FieldStorageConfig;
+use Drupal\rules\Context\ContextConfig;
+use Drupal\rules\Event\EntityEvent;
+
+/**
+ * Tests events with qualified name.
+ *
+ * @group rules
+ */
+class EventHandlerTest extends RulesDrupalTestBase {
```
Comment (fago) [inline tests/src/Kernel/EventHandlerTest.php]: Should be ConfigurableEventHandlerTest

File: tests/src/Kernel/EventHandlerTest.php [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+      ]);
+  }
+
+  /**
+   * Tests ConfigurableEventHandlerEntityBundle.
+   *
+   * @todo Add integrity check that node.field_integer is detected by Rules.
+   */
+  public function testEntityBundleHandler() {
+    // Create rule1 with the 'rules_entity_presave:node--page' event.
```
Comment (fago) [inline tests/src/Kernel/EventHandlerTest.php]: would should add some high level summary of we are doing: We test that rules are triggered correctly based upon the fully qualified event name as well as the base event name.

File: src/Entity/ReactionRuleStorage.php [hunk 28]
```diff
@@ -61,6 +69,19 @@ public function __construct(EntityTypeInterface $entity_type, ConfigFactoryInter
...
 
   /**
+   * Gets the event manager.
+   *
+   * @return \Drupal\rules\Core\RulesEventManager
+   *   The event manager.
+   */
+  protected function eventManager() {
+    if (!$this->eventManager) {
+      $this->eventManager = new RulesEventManager($this->moduleHandler());
```
Comment (fago) [inline src/Entity/ReactionRuleStorage.php]: nope - this must use either proper depdency injection or the pattern as e.g. \Drupal\Core\Form\FormBase::getRouteMatch. As the class already uses proper constructor injection, we should use that here also. In addition, this should be getEventManager().

File: src/EventHandler/ConfigurableEventHandlerEntityBundle.php [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+      $events_suffixes[] = $event->getSubject()->bundle();
+    }
+    return $events_suffixes;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function summary() {
+    // Nothing to do by default.
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php:33]: missing implementation - check the summary provided by the d7 implementation and follow the same here

File: src/EventHandler/ConfigurableEventHandlerEntityBundle.php [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+   */
+  public function validate() {
+    // Nothing to check by default.
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getEventNameSuffix() {
+    // Nothing to do by default.
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php:54]: Missing implementation

File: src/EventHandler/ConfigurableEventHandlerEntityBundle.php [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function refineContextDefinitions() {
+    // Nothing to refine by default.
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php]: can be in base class

File: src/EventHandler/ConfigurableEventHandlerEntityBundle.php [hunk 32]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function calculateDependencies() {
+    // Nothing to calculate by default.
```
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php]: can be in base class

File: src/Entity/ReactionRuleStorage.php (hunk 28 above)
Comment (milkovsky) [inline src/Entity/ReactionRuleStorage.php]: there is no \Drupal::eventManager(); and not $container->get('event_manager').
Is event_manager a core thing?
I cannot find an example  where it is initialized

Comment (milkovsky) [inline src/Entity/ReactionRuleStorage.php]: Found . it is $container->get('plugin.manager.rules_event')

File: src/EventHandler/ConfigurableEventHandlerEntityBundle.php (hunk 29 above)
Comment (milkovsky) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php:33]: Do you think we should store event_name in configuration?

in d7 entityInfo is stored in $entityInfo property. looks strange Should I do the same?

You offered to extend Drupal\Component\Plugin\PluginBase in ConfigurableEventHandlerEntityBundle

Coustruct has:
function __construct(array $configuration, $plugin_id, $plugin_definition) {
    parent::__construct($configuration, $plugin_id, $plugin_definition);
  }
what should I put in there? There should be $event and configuration instead?

File: src/EventHandler/ConfigurableEventHandlerEntityBundle.php (hunk 30 above)
Comment (milkovsky) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php:54]: Isnt it duplicate of ConfigurableEventHandlerEntityBundle::determineQualifiedEvents() ?

Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php:54]: almost, but as documented in the class docs:
- The event name suffix must be generated from the event data at run-time,
- while the configured plugin has to determine it based upon the event
- configuration.

So we need both.

File: src/EventHandler/ConfigurableEventHandlerEntityBundle.php (hunk 29 above)
Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php:33]: $plugin_id is the (base) event_name. configuration is configuration.

> Do you think we should store event_name in configuration?

No, storage is fine as we have it now afaics.

Comment (milkovsky) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php:33]: > the (base) event_name

Q1 Would be more useful to have event object instead?

> plugin_definition

Q2 left unanswered

> in d7 entityInfo is stored in $entityInfo property. looks strange Should I do the same?

Q3 left unanswered

File: src/EventHandler/ConfigurableEventHandlerEntityBundle.php (hunk 30 above)
Comment (milkovsky) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php:54]: We should have event object to do it. Is it stored in the class ?

Class is not instantiated anywhere. It is hard to write the blind code. We should postpone it once the methods are needed/used.

Comment (fago) [inline src/EventHandler/ConfigurableEventHandlerEntityBundle.php:54]: right, that part comes later with the UI - so ok, let's postpone this part.

**Comment by fago** [comment]
Merged.

**Pull request closed without merging**